lib/pre-build.js

```diff
@@ -0,0 +1,367 @@
+    } else {
+      cb(null)
+    }
+  })
+}
+
+function build(target, rebuild, cb) {
+  if (IS_WIN) {
+    target = '/t:' + target
+  }
+
+  var cmds = rebuild ? ['clean', 'configure'] : ['configure']
+
+  execGyp(cmds, function cleanCb(err) {
+    if (err) {
+      return cb(err)
+    }
+
+    var jobs = Math.round(CPU_COUNT / 2)
+    execGyp(['build', '-j', jobs, target], cb)
+  })
+}
+
+function moveBuild(target, cb) {
+  var filePath = path.join(BUILD_PATH, target + '.node')
+  var destination = path.join(BUILD_PATH, getBinFileName())
+  fs.rename(filePath, destination, cb)
+}
+
+function download(cb) {
+  var hasCalledBack = false
+  var fileName = getPackageFileName()
+  var url = DOWNLOAD_HOST + REMOTE_PATH + fileName
+  https.get(url, function getFile(res) {
+    if (res.statusCode !== 200) {
+      return cb(new Error('Failed to download ' + url + ': code ' + res.statusCode))
+    }
+
+    var unzip = zlib.createGunzip()
+    var buffers = []
+    var size = 0
+    res.pipe(unzip).on('data', function onResData(data) {
+      buffers.push(data)
+      size += data.length
+    })
+
+    res.on('error', function onResError(err) {
+      if (!hasCalledBack) {
+        hasCalledBack = true
+        cb(new Error('Failed to download ' + url + ': ' + err))
+      }
+    })
+
+    unzip.on('error', function onResError(err) {
+      if (!hasCalledBack) {
+        hasCalledBack = true
+        cb(new Error('Failed to unzip ' + url + ': ' + err))
+      }
+    })
+
+    unzip.on('end', function onResEnd() {
+      if (hasCalledBack) {
+        return
+      }
+      hasCalledBack = true
+      cb(null, Buffer.concat(buffers, size))
+    })
+
+    res.resume()
+  })
+}
+
+function saveDownload(data, cb) {
+  makePath(BUILD_PATH, function makePathCB(err) {
+    if (err) {
+      return cb(err)
+    }
+
+    var filePath = path.join(BUILD_PATH, getBinFileName())
+    fs.writeFile(filePath, data, cb)
+  })
+}
+
+function install(target, cb) {
+  // First, attempt to build the package using the source. If that fails, try
+  // downloading the package. If that also fails, whoops!
+  var errors = []
+  build(target, true, function buildCB(err) {
+    if (!err) {
+      return moveBuild(target, function moveBuildCB(err) {
+        if (err) {
+          errors.push(err)
+          doDownload()
+        } else {
+          doCallback()
+        }
+      })
+    }
+    errors.push(err)
+
+    // Building failed, try downloading.
+    doDownload()
+  })
+
+  function doDownload() {
+    if (opts['no-download']) {
+      return doCallback(new Error('Downloading is disabled.'))
+    }
+
+    download(function downloadCB(err, data) {
+      if (err) {
+        return doCallback(err)
+      }
+
+      saveDownload(data, doCallback)
+    })
+  }
+
+  function doCallback(err) {
+    if (err) {
+      errors.push(err)
+    }
+
+    if (errors.length > 0) {
+      cb(new Error('Failed to install module: ' + errors.join('; ')))
```

Comment:
Doh, there's a flaw here. Shouldn't consider this a failure just because some step failed if the last step did not fail.